rsmtool/rsmpredict.py

```diff
@@ -88,41 +93,68 @@ def fast_predict(
 
         This dataframe should be read from the "feature.csv" file under the
         "output" directory of the previously run RSMTool experiment.
+
+        Defaults to ``None``.
+    trim : bool
+        Whether to trim the predictions. If ``True``, ``trim_min`` and
+        ``trim_max`` must be specified or be available as attributes of
+        the ``modeler``.
+        Defaults to ``False``.
     trim_min : int, optional
         The lowest possible integer score that the machine should predict.
-        If ``None``, raw predictions will not be trimmed.
+        If ``None``, this function will try to extract this value from
+        ``modeler``. If ``None`` and no such attribute exists, raw
```

Comment:
_"If ``None`` and no such attribute exists, raw predictions will not be trimmed."_

if `trim` is set to True, then won't this scenario result in a TypeError [at this part of preprocessor.trim](https://github.com/EducationalTestingService/rsmtool/blob/4c5cab6b20f845426f34998dc652241a3fe2ebd5/rsmtool/preprocessor.py#L413-L414)?

Reply:
It would have led to an AttributeError, I believe. However, I have converted it to a ValueError now with a helpful message. Of course, if the attribute or what you pass in is not a float, then, yes, I think a TypeError would be thrown. I'm not sure it's worth it to validate to that level.

Reply:
@mulhod It doesn't give an error like this when trying to add the max_trim to the tolerance?
TypeError: unsupported operand type(s) for +: 'NoneType' and 'float'

Reply:
This merged now, but I wanted to show you what happens when, say, `trim_max` doesn't exist as an attribute. It's caught before it ever gets to the point where `trim_max` would be added to `trim_tolerance`.

```shell
❯ ../env/bin/ipython
Python 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from rsmtool.rsmpredict import fast_predict

In [2]: from rsmtool.modeler import Modeler

In [3]: m = Modeler.load_from_file("tests/data/experiments/lr-predict/existing_experiment/output/lr.model")

In [4]: features = {
   ...:             "FEATURE1": 6.0,
   ...:             "FEATURE2": 1.0,
   ...:             "FEATURE3": -0.2,
   ...:             "FEATURE4": 0,
   ...:             "FEATURE5": -0.1,
   ...:             "FEATURE6": 5.0,
   ...:             "FEATURE7": 12,
   ...:             "FEATURE8": -7000,
   ...:         }

In [5]: delattr(m, "trim_max")

In [6]: fast_predict(features, m, trim=True)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line 1
----> 1 fast_predict(features, m, trim=True)

File ~/Documents/rsmtool/rsmtool/rsmtool/rsmpredict.py:258, in fast_predict(input_features, modeler, df_feature_info, trim, trim_min, trim_max, trim_tolerance, scale, train_predictions_mean, train_predictions_sd, h1_mean, h1_sd, logger)
    256         trim_max = modeler.trim_max
    257 except AttributeError:
--> 258     raise ValueError(
    259         "The trim_min/trim_max modeler attributes must exist when trim=True "
    260         "and these values are not explicitly specified"
    261     ) from None
    262 for column in df_predictions.columns:
    263     df_predictions[f"{column}_trim"] = preprocessor.trim(
    264         df_predictions[column], trim_min, trim_max, trim_tolerance
    265     )

ValueError: The trim_min/trim_max modeler attributes must exist when trim=True and these values are not explicitly specified
```

Now, If we specifically pass in a value for `trim_max`, in which case it will not check for the attribute, and that value is not a number of some kind, then it will result in a `TypeError`. But, this is expected since the value passed in is not a valid `trim_max`. This could be caught and validated, but 1) it's a strange thing to do and 2) the error message is not unclear. This is why I don't think it's worth it to validate. It would still be an error, so it's not like it would silently fail.

```shell
In [7]: fast_predict(features, m, trim=True, trim_max="not_a_number")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[10], line 1
----> 1 fast_predict(features, m, trim=True, trim_max="not_a_number")

File ~/Documents/rsmtool/rsmtool/rsmtool/rsmpredict.py:263, in fast_predict(input_features, modeler, df_feature_info, trim, trim_min, trim_max, trim_tolerance, scale, train_predictions_mean, train_predictions_sd, h1_mean, h1_sd, logger)
    258         raise ValueError(
    259             "The trim_min/trim_max modeler attributes must exist when trim=True "
    260             "and these values are not explicitly specified"
    261         ) from None
    262     for column in df_predictions.columns:
--> 263         df_predictions[f"{column}_trim"] = preprocessor.trim(
    264             df_predictions[column], trim_min, trim_max, trim_tolerance
    265         )
    266         df_predictions[f"{column}_trim_round"] = np.rint(
    267             df_predictions[f"{column}_trim"]
    268         ).astype("int64")
    269 elif any(arg is not None for arg in [trim_tolerance, trim_min, trim_max]):

File ~/Documents/rsmtool/rsmtool/rsmtool/preprocessor.py:413, in FeaturePreprocessor.trim(self, values, trim_min, trim_max, tolerance)
    410 if isinstance(values, list):
    411     values = np.array(values)
--> 413 new_max = trim_max + tolerance
    414 new_min = trim_min - tolerance
    415 trimmed_values = values.copy()

TypeError: can only concatenate str (not "float") to str
```

Reply:
@mulhod But in normal usage the trim_min/max attributes are set to None when the model is instantiated, right?. I think you're only getting the attribute error by artificially deleting the model attribute at In [5] and then later trying to set the nonexistent variable. With the variable in existence as None I think it would fall into that second case you demonstrated.

Reply:
In the standard usage line 256 (i.e., trim_max = modeler.trim_max) would just be setting trim_max to None, right? Am I  still missing something here? Either way, thanks for taking the time to discuss this further even after it's been merged!

Reply:
Perhaps I don't understand what you mean. Could you demonstrate it with an example?

Reply:
I don't see any reason in the code that the modeler.trim_max instance wouldn't exist. Even if you're working with a skll model, load_from_learner() still instantiates a new Modeler instance and just sets the learner attribute. So the other class attributes should still exists and are set to None

Reply:
You're right that instantiating a modeler instance will have those attributes set to `None`, but those instances are only saved as part of running `rsmtool`, in which case the attributes would be there. Basically, we don't typically instantiate an instance anywhere outside of `rsmtool`, so modeler objects are not saved by anything other than `rsmtool`. The only case where the `TypeError` would occur is if you instantiated a modeler manually and then just saved it without really doing anything else. It's the kind of edge case that I'm not sure we need to consider right now. Even if you were doing something like this, you would have to have good knowledge about the internal workings of these things, so it probably wouldn't need to be caught.

Reply:
> The only case where the TypeError would occur is if you instantiated a modeler manually and then just saved it without really doing anything else

A much more common occurrence would be using  load_from_learner() with a model that was built/saved with the previous version of RSMTool and the result would be the same, wouldn't it? Either way I agree that this isn't worth spending any more time on since it is just an edge case. 

Although, @mulhod Do you think there could be any argument for adding a model version attribute to the newer rsmtool model objects? Then when load_from_learner() is used, it could automatically display the version of the model to help distinguish between older models and newer?

Reply:
@blongwill I hope this answers your question. Below I load in an old model and try to make a prediction with it. The attribute is not None, it simply doesn't exist. It throws the expected error.

```shell
# Check out old model
❯ git checkout v9.1.1 tests/data/experiments/lr-predict/existing_experiment/output/lr.model
❯ ../env/bin/ipython
Python 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.14.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from rsmtool.modeler import Modeler

In [2]: from rsmtool.rsmpredict import fast_predict

In [3]: m = Modeler.load_from_file("tests/data/experiments/lr-predict/existing_experiment/output/lr.model")
Trying to unpickle estimator StandardScaler from version 1.2.0 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations
Trying to unpickle estimator LinearRegression from version 1.2.0 when using version 1.3.0. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/model_persistence.html#security-maintainability-limitations

In [4]: m.feature_info
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 m.feature_info

AttributeError: 'Modeler' object has no attribute 'feature_info'

In [5]: features = {
   ...:             "FEATURE1": 6.0,
   ...:             "FEATURE2": 1.0,
   ...:             "FEATURE3": -0.2,
   ...:             "FEATURE4": 0,
   ...:             "FEATURE5": -0.1,
   ...:             "FEATURE6": 5.0,
   ...:             "FEATURE7": 12,
   ...:             "FEATURE8": -7000,
   ...:         }

In [6]: fast_predict(features, m)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line 1
----> 1 fast_predict(features, m)

File ~/Documents/rsmtool/rsmtool/rsmtool/rsmpredict.py:197, in fast_predict(input_features, modeler, df_feature_info, trim, trim_min, trim_max, trim_tolerance, scale, train_predictions_mean, train_predictions_sd, h1_mean, h1_sd, logger)
    195         df_feature_info = modeler.feature_info
    196     except AttributeError:
--> 197         raise ValueError(
    198             "df_feature_info must be specified if it cannot be found as an "
    199             "attribute in the modeler object"
    200         ) from None
    202 # preprocess the input features so that they match what the model expects
    203 try:

ValueError: df_feature_info must be specified if it cannot be found as an attribute in the modeler object
```

Can you tell me what about this might be unexpected or problematic?

Reply:
Ok, I think I may understand what you're getting at now. When loading from a bare SKLL model file, the `trim_min`, etc., attributes will actually exist since they are set to `None`. That is, they won't simply be missing, which my tests were assuming. They will be there and have a value of `None`. So, this would be a problem. It's not a very big problem because they would encounter an exception eventually (a `TypeError`, as you mention). But, it's nice to catch this earlier perhaps so that it's clear what the problem is. I will make a merge request to show what I mean.